### DISCUSSION TITLE
Issue with unicode string processing

### DISCUSSION THREAD

**Discussion by XiaoQing235** [cc-tweaked/CC-Tweaked#2035] [Q&A] [answered]
I'm using a computer to control `Create_DisplayLink` and need to include Chinese characters. However, writing them directly didn’t work. I tried using Java escape sequences to send Unicode characters to create_displaylink, but they end up being capitalized. I've gone through all the related issues, pull requests, and discussions about Unicode, UTF-8, and CJK, but it seems like there's no solution.

heres are my scripts.
main.lua
```
local modem = peripheral.find("modem") or error("No modem attached", 0)
local inductionPort = peripheral.find("inductionPort") or error("No Induction Port found.", 0)
local monitor = peripheral.find("monitor") or error("No Monitor found.", 0)

term.redirect(monitor)
monitor.setTextScale(1)

modem.open(1)

while true do
    local energy = inductionPort.getEnergy() *0.4 / 1000000
    local maxEnergy = inductionPort.getMaxEnergy() *0.4 / 1000000
    local percent = (energy / maxEnergy) * 100

    monitor.clear()
    monitor.setCursorPos(1, 1)

    local text = string.format("\\u80fd\\u91cf\\uff1a%.2f MFE / %.2f MFE    %.2f%%", energy, maxEnergy, percent)
    print(text)
    modem.transmit(1, 1, text)
    sleep(0.25)
end
```
display.lua
```
local link = peripheral.find("Create_DisplayLink") or error("No Create Display Link attached", 0)
local monitor = peripheral.find("monitor") or error("No Monitor found.", 0)
local modem = peripheral.find("modem") or error("No modem attached", 0)
modem.open(1)
term.redirect(monitor)
monitor.setTextScale(1)

function displayText(text)
    link.setCursorPos(1,1)
    link.clearLine()
    link.write(text)
    link.update()
end

while true do
    local event, side, channel, replyChannel, message, distance = os.pullEvent("modem_message")
    displayText(message)
    monitor.clear()
    monitor.setCursorPos(1, 1)
    print(message)
end
```
I don't need to display these characters on the computercraft monitor, just send them correctly to `Create_DisplayLink`

This is my first time posting a github discussion, so please correct me if I've done something wrong!

**Comment by zyxkad** [comment]
In CC:T, string only stores as byte array inside the [cobalt] lua machine. At anytime it goes to the peripherals, the string will be represented as char sequences, which will *not* be encoded as utf8 or any other char types in the operation.
So if you want to present any non ascii character, it will need some support on the addon mod to allow you display non ascii using ascii sequences. Or wait till SquidDev decided to support utf8 on cobalt.

https://github.com/cc-tweaked/Cobalt/blob/b2c0489d550ff416596579adac83492276794afc/src/main/java/org/squiddev/cobalt/LuaString.java#L520

**Comment by Wojbie** [comment, marked as answer]
You can see related discussion about it in #860 but at current time there really is no feasible way for CC to display unicode characters internally.

---
Alternative solution i have seen players use is to write custom font texture font for CC:T basically replacing upper rows (from A_ onward) 
![obraz](https://github.com/user-attachments/assets/890d8f14-6845-4d68-b720-649445f8d447)
With their own custom characters, and then writing programs that display those characters, knowing they are replaced in texture file with what user wants. This is sadly not universal solution as it requires all players to use modified texture pack to see modified characters and requires writing a program in quite annoying way to display them.

---
If you had a peripheral that was able to understand utf8 endcoded string then CC:T has backported lua 5.3 utf8 library <https://www.lua.org/manual/5.3/manual.html#6.5> so one could easily create encoded string but i don't believe that any current peripheral mods that display text offer that option.
